src/core/storage.rs

```diff
@@ -206,27 +230,213 @@ impl<K: Hash + Clone + Eq + Debug> RecordsStorage<K> {
-		Q: Hash + Eq,
 	{
 		self.direct_records.get(key).cloned()
 	}
 
-	/// Size of the storage
-	pub fn len(&self) -> usize {
+	fn len(&self) -> usize {
 		self.direct_records.len()
 	}
 
-	/// Returns all keys in the storage
-	pub fn keys(&self) -> Vec<K> {
+	fn keys(&self) -> Vec<K> {
 		self.direct_records.keys().cloned().collect()
 	}
 }
 
+/// This trait is used to define a storage that can store items organised in prefixes.
+/// Prefixes are used to group elements by some characteristic. For example, to get
+/// elements that belong to some particular parachain.
+pub trait PrefixedRecordsStorage<K, P> {
+	/// Insert a prefixed entry to the storage
+	fn insert_prefix(&mut self, prefix: P, key: K, entry: StorageEntry) -> color_eyre::Result<()>;
+	/// Replaces a prefixed entry in the storage, both prefix and a key must exist
+	fn replace_prefix<Q: ?Sized + Hash + Eq, PQ: ?Sized + Hash + Eq>(
+		&mut self,
+		prefix: &PQ,
+		key: &Q,
+		entry: StorageEntry,
+	) -> Option<StorageEntry>
+	where
+		K: Borrow<Q>,
+		P: Borrow<PQ>;
+	/// Get a key using specific prefix along with the key
+	fn get_prefix<Q: ?Sized + Hash + Eq, PQ: ?Sized + Hash + Eq>(&self, prefix: &PQ, key: &Q) -> Option<StorageEntry>
+	where
+		K: Borrow<Q>,
+		P: Borrow<PQ>;
+	/// Get keys for a specific prefix
+	fn prefixed_keys<PQ: ?Sized + Hash + Eq>(&self, prefix: &PQ) -> Vec<K>
+	where
+		P: Borrow<PQ>;
+	/// Get all prefixes from a storage
+	fn prefixes(&self) -> Vec<P>;
+}
+
+/// Prefixed storage is distinct as it organise data stored using prefixes,
+/// for example to store entries for different parachains and relay parents
+/// The keys should be unique in all distinct prefixes, that can be
+/// guaranteed by assuming that K is a cryptographic hash
+/// This data structure is intended to work with a small and limited number of
+/// prefixes, as it will likely perform a hash lookup per each prefix
+/// when searching for a key in non-prefixed matter
+pub struct HashedPrefixedRecordsStorage<K: Hash + Clone, P: Hash + Clone> {
+	/// The configuration.
+	config: RecordsStorageConfig,
+	/// The last block number we've seen. Used to index the storage of all entries.
+	last_block: Option<BlockNumber>,
+	/// Elements with expire dates.
+	ephemeral_records: BTreeMap<BlockNumber, HashSet<K>>,
+	/// Direct mapping to values.
+	prefixed_records: HashMap<P, HashMap<K, StorageEntry>>,
+}
+
+impl<K, P> RecordsStorage<K> for HashedPrefixedRecordsStorage<K, P>
+where
+	K: Hash + Clone + Eq + Debug,
+	P: Hash + Clone + Eq + Debug,
+{
+	fn new(config: RecordsStorageConfig) -> Self {
+		let ephemeral_records = BTreeMap::new();
+		let prefixed_records = HashMap::new();
+		Self { config, last_block: None, ephemeral_records, prefixed_records }
+	}
+
+	// We cannot insert non prefixed key into a prefixed storage
+	fn insert(&mut self, key: K, _: StorageEntry) -> color_eyre::Result<()> {
+		return Err(eyre!("trying to insert key with no prefix to the prefixed storage: {:?}", key))
```

Comment:
We could also also allow this operation and use a default prefix. It is more flexible and would allow old written code to still work with a prefixed storage, even if it's code doesn't actually need prefixes.

Reply:
One problem I've found is to define what is a default prefix for type `P`? We could use `Default` trait for that but it adds an additional burden on `P` type...